app/src/main/java/com/keylesspalace/tusky/util/LinkHelper.kt

```diff
@@ -0,0 +1,251 @@
+
+    for (span in urlSpans) {
+        setClickableText(span, builder, mentions, tags, listener)
+    }
+    view.text = builder
+    view.movementMethod = LinkMovementMethod.getInstance()
+}
+
+@VisibleForTesting
+fun setClickableText(
+    span: URLSpan,
+    builder: SpannableStringBuilder,
+    mentions: List<Mention>?,
+    tags: List<HashTag>?,
+    listener: LinkListener
+) {
+    val start = builder.getSpanStart(span)
+    val end = builder.getSpanEnd(span)
+    val flags = builder.getSpanFlags(span)
+    val text = builder.subSequence(start, end)
+
+    val customSpan = when {
+        text[0] == '#' -> getCustomSpanForTag(tags, span, listener)
+        text[0] == '@' -> getCustomSpanForMention(mentions, span, listener)
+        else -> null
+    } ?: object : NoUnderlineURLSpan(span.url) {
+        override fun onClick(view: View) {
+            listener.onViewUrl(url)
+        }
+    }
+
+    builder.removeSpan(span)
+    builder.setSpan(customSpan, start, end, flags)
+
+    /* Add zero-width space after links in end of line to fix its too large hitbox.
+     * See also : https://github.com/tuskyapp/Tusky/issues/846
+     *            https://github.com/tuskyapp/Tusky/pull/916 */
+    if (end >= builder.length || builder.subSequence(end, end + 1).toString() == "\n") {
+        builder.insert(end, "\u200B")
+    }
+}
+
+private fun getCustomSpanForTag(tags: List<HashTag>?, span: URLSpan, listener: LinkListener): ClickableSpan? {
+    return when (val tag = tags?.firstOrNull { it.url == span.url }) {
+        null -> null
+        else -> object : NoUnderlineURLSpan(span.url) {
+            override fun onClick(view: View) {
+                listener.onViewTag(tag.name)
+            }
+        }
+    }
+}
+
+private fun getCustomSpanForMention(mentions: List<Mention>?, span: URLSpan, listener: LinkListener): ClickableSpan? {
+    // https://github.com/tuskyapp/Tusky/pull/2339
+    return when (val mention = mentions?.firstOrNull { it.url == span.url }) {
+        null -> null
+        else -> getCustomSpanForMentionUrl(span.url, mention.id, listener)
+    }
+}
+
+private fun getCustomSpanForMentionUrl(url: String, mentionId: String, listener: LinkListener): ClickableSpan {
+    return object : NoUnderlineURLSpan(url) {
+        override fun onClick(view: View) {
+            listener.onViewAccount(mentionId)
+        }
+    }
+}
+
+/**
+ * Put mentions in a piece of text and makes them clickable, associating them with callbacks to
+ * notify when they're clicked.
+ *
+ * @param view the returned text will be put in
+ * @param mentions any '@' mentions which are known to be in the content
+ * @param listener to notify about particular spans that are clicked
+ */
+fun setClickableMentions(view: TextView, mentions: List<Mention>?, listener: LinkListener) {
+    if (mentions?.isEmpty() != false) {
+        view.text = null
+        return
+    }
+
+    val builder = SpannableStringBuilder()
```

Comment:
There is a `buildSpannedString` top level function that would make this function a little bit easier to write https://developer.android.com/reference/kotlin/androidx/core/text/package-summary#buildSpannedString(kotlin.Function1)

It kinda sucks that it converts the result to `SpannedString` which is an additional object allocation we don't need though